File: web.py
#!/usr/bin/env python

weblog = '/var/log/httpd/access_log'
file = open( weblog, 'r' )

ips = set()

#
# build a set of unique IP addresses
#
def get_unique_ips():
    for line in file:
        column = line.split(' ')
        ip = column[0]
        ips.add( ip )

get_unique_ips()

#
# get a count of how many times the IP shows up in the logfile
#
def get_ip_count( address ):
    count = 0
    f = open( weblog, 'r' )
    for line in f:
        if address in line:
            count += 1
    return count

ip_last_dates = set()

#
# get the last date string from the visiting IP address
#
def get_last_dates( date_count ):
    count = 0
    f = open( weblog, 'r' )
    for line in f:
        if address in line:
            count += 1
            if count == date_count:
                column = line.split(' ')
                date = column[2] + " " + column[3] + " " + column[4]
                return date

ip_counts = set()

for address in ips:
    ip_counts.add( get_ip_count( address ) )
    count = get_ip_count( address )
    ip_last_dates.add( get_last_dates( count ))

logs = set()

print "IP Addr | Num visits | Date of last visit"
for date in ip_last_dates:
    f = open( weblog, 'r' )
    for line in f:
        column = line.split(' ')
        ip = column[0]
        ip_count = get_ip_count( ip )
        str_ip_count = str( ip_count )
        if date in line:
            log = ip + " --  " + str_ip_count + " --  " + column[2] + " " + column[3] + " " + column[4]
            logs.add( log.rstrip() )

for entry in logs:
    print entry

File: web.pl
#!/usr/bin/perl

use strict;
use warnings;
use autodie;
use Socket;

my @ips;
my $access_log = "/var/log/httpd/access_log";

sub get_iplist {
    open my $fh, "<", $access_log;
    while ( <$fh> ) {
        my @line = split ' ', $_;
        push @ips, $line[0];
    }
    close $fh;
}

sub main {
    get_iplist();
    my %ips_counts = ();
    for my $ip ( @ips ) {
        my $count = grep( /$ip/, @ips ); 
        $ips_counts{ $ip } = $count unless exists $ips_counts{ $ip };
    }

    foreach my $key ( sort { $ips_counts{ $b } <=> $ips_counts{ $a } } keys %ips_counts ) {
        my $host = gethostbyaddr( inet_aton( $key ), AF_INET );
        if ( $ips_counts{ $key } and $key and $host ) {
            printf( "%-10s - %-20s - %s\n", $ips_counts{ $key }, $key, $host );
        }
    }
}
